# Let me be clear man

We have worked we the **arrays** a lot. I mean a lot by this time.<br>

1. Firstly **introduced** in the C
2. Then worked a lot in it **producing many sorting algorithms** (remember the **duplicate deletion** algorithm you have made?)
3. Then skipping time (java, php etc.) was introduced in **python** in the form of lists
4. Then in the final year **project** (numpy)
5. Then skipping... read the **book** on that!
6. Numpy! Array! Friends!
7. **Series**, Pandas dataframe... all arrays!
8. And now, **here**.

Mhuum.. so... arrays.

# 

### Do not forget

In general python `list != array` of course... I am wasting my time here. Let's get to the main point.

**The Sequences Datatypes**: <br>
- List
- Tuple
- String
- Byte Array
- Range()

These ↑ are the main sequence datatypes. 

# 

### The memory

Here, we just need to revise some already known topics. So, just taking notes from the lecture.

- The basic storage type is bytes (actually bits)
- In RAM the access of any single individuall memory block is $O(1)$
- For python the **single character** is the UNICODE which takes 16bits / 2bytes.
- Python (list, tuples) are the **referantial structures** as they store the "references" instead of the actual object.

# 

### Simply saves the references.

<img src="../ref_imgs/mem_1.png" height=300 width=500>

### Just change the reference! Not data (if immutable!)

<img src="../ref_imgs/mem_2.png" height=300 width=500>

### Don't make mistake again ↓
Whenever there is the syntax like `a = [1, 2, 44] * 3` then it means that ↓

<img src="../ref_imgs/mem_3.png" height=300 width=500>

Refer to the same object.

- `b = list(a)` → Shallow Copy: Will just assign the same references to the new variable
- `b = deepcopy(a)` → Deep Copy: Recreate the internal objects and then assign to the new variable

**NOTE**: The both of these ↑ are *not* applicable for the "immutable" types. Both will return the same result.

# 

# Dynamic Arrays

These are the arrays which are not fixed in size. They grow in memory are needed. In python we have seen that "list" is that kind of "equential" data structure which can be called "Dynamic" in nature.

The other ones are the Dictionary, Set but they are **not** array to say (not sequential)

___

Iterastingly, for the first time I have seen that, "not for each append" the lists allocate memory. But it "rents in advance". Meaning ↓

In [174]:
n = 20
list_ = []

for i in range(n):
    print(f"The len is: {len(list_):4d} & the size is: {list_.__sizeof__():4d}")
    list_.append(i)

The len is:    0 & the size is:   40
The len is:    1 & the size is:   72
The len is:    2 & the size is:   72
The len is:    3 & the size is:   72
The len is:    4 & the size is:   72
The len is:    5 & the size is:  104
The len is:    6 & the size is:  104
The len is:    7 & the size is:  104
The len is:    8 & the size is:  104
The len is:    9 & the size is:  168
The len is:   10 & the size is:  168
The len is:   11 & the size is:  168
The len is:   12 & the size is:  168
The len is:   13 & the size is:  168
The len is:   14 & the size is:  168
The len is:   15 & the size is:  168
The len is:   16 & the size is:  168
The len is:   17 & the size is:  232
The len is:   18 & the size is:  232
The len is:   19 & the size is:  232


See that, it started **initially** with taking `40` bytes and then when the first element was added (2nd line) it took directly 72. And kept 72 for 4 elements. Which means (72 - 40) / 4 = `8 bytes` per reference.

And the same goes for `168` bytes when it reserved 8 blocks for future instead of 4. *(168 - 104) / 8 = `8 bytes`*

**Key takeaway**: Is that, the list grows, but not for each element. It grows for multiple **potential** elements. So even if you have 3 elements in the list in total, the space it reserves in the memory is for 5 element references.

# 

## For the first time... 

...and also as for the trial purposes, we will be building the "Dynamic Arrays" or Lists in python from *kind of* scratch. Here, we will discuss the simple steps:

1. Allocate the memory for the initial array and cally it `ary`
2. Now, we want to grow the array, but **we can't actually grow** the array, so we will re initialize it.
3. Create a new ary and call it `ary_2` with larger capacity than `ary`. *(just like more capacity the list does we saw earlier)*
4. Then `ary_2 = ary` and then re-initialize it with `ary = ary_2`.

###### 

<img src="../ref_imgs/ary_grow.png" height=300 width=500>

###### 

# But,
That is the story for the next book! See you there!

# 

Haha! Here is our new buddy with whom I am typing for the first time! Have a look ↓ <br>
*(Date: 21 Feb, 2022)*

<img src="../ref_imgs/keybee.jpg" width=600 height=400>